README
NAME
    App::bmkpasswd - bcrypt-capable mkpasswd(1) and exported helpers

SYNOPSIS
      bmkpasswd --help
  
      ## Generate bcrypted passwords
      ## Defaults to work cost factor '08':
      bmkpasswd
      bmkpasswd --workcost='06'

      ## Use other methods:
      bmkpasswd --method='md5'
      # SHA requires Crypt::Passwd::XS or glibc2.7+
      bmkpasswd --method='sha512'
  
      ## Compare a hash:
      bmkpasswd --check=HASH

      ## Check hash generation times:
      bmkpasswd --benchmark

DESCRIPTION
    App::bmkpasswd is a simple bcrypt-enabled mkpasswd. (Helper functions
    are also exported for use in other applications; see "EXPORTED".)

    See "bmkpasswd --help" for usage information.

    Uses Crypt::Eksblowfish::Bcrypt for bcrypted passwords. Bcrypt hashes
    come with a configurable work-cost factor; that allows hash generation
    to become configurably slower as computers get faster, thereby impeding
    brute-force hash generation attempts.

    See <http://codahale.com/how-to-safely-store-a-password/> for more on
    why you ought to be using bcrypt or similar "adaptive" techniques.

    SHA-256 and SHA-512 are supported if available. You'll need either
    Crypt::Passwd::XS or a system crypt() that can handle SHA, such as
    glibc-2.7+ or newer FreeBSD builds.

    MD5 support is fairly universal, but it is known insecure and there is
    really no valid excuse to be using it; it is included here for
    compatibility with ancient hashes.

    Salts are randomly generated.

EXPORTED
    You can use the exported mkpasswd and passwdcmp functions in other Perl
    modules/applications:

      use App::bmkpasswd qw/mkpasswd passwdcmp/;

      ## Generate a bcrypted passwd with work-cost 08:
      $bcrypted = mkpasswd($passwd);

      ## Generate a bcrypted passwd with other work-cost:
      $bcrypted = mkpasswd($passwd, 'bcrypt', '06');

      ## SHA:
      $crypted = mkpasswd($passwd, 'sha256');
      $crypted = mkpasswd($passwd, 'sha512');

      ## Compare a password against a hash
      ## passwdcmp() will return the hash if it is a match
      if ( passwdcmp($passwd, $hash) ) {
        ## Successful match
      } else {
        ## Failed match
      }

BUGS
    There is currently no easy way to pass your own salt; frankly, this
    thing is aimed at some projects of mine where that issue is unlikely to
    come up and randomized is appropriate. If that's a problem, patches
    welcome? ;-)

AUTHOR
    Jon Portnoy <avenj@cobaltirc.org>



Changes
Revision history for Perl module App::bmkpasswd

2.012001 2016-07-25

  - Safer Crypt::Passwd::XS require() (ignore '.' in @INC)

  - bmkpasswd executable always drops '.' from @INC

  - Default salt generator now package-scoped

  - Test suite improvements

2.011006 2016-03-20

  - Improved check for POSIX::Termios in `bmkpasswd`

  - Minor `bmkpasswd` cleanups; take advantage of IO::Handle methods as long
    as we have it

2.011005 2016-03-09

  - Skip 'bmkpasswd' command tests on MSWin32

    While the utility is reported to work without issue, the test does not,
    and I lack access to an appropriate machine for deeper diving.

2.011004 2016-03-09

  - Fall back to Term::ReadKey on MSWin32

2.011003 2016-03-09

  - Fall back to Term::ReadKey to turn off terminal echo on 5.8.x, if
    available; if not, warn loudly and suggest it before prompting for a
    password to crypt

2.011002 2016-03-09

  - Fix `bmkpasswd` executable on 5.8.x (no POSIX::ECHO available)

2.011001 2016-03-08

  - Add `bmkpasswd --available` for listing available crypt methods

  - Adjust constant time comparison to avoid potentially leaking any
    information regarding the length of random-length SHA salts

  - Add proper Test::Cmd tests for `bmkpasswd`

2.010001 2015-10-23

  - Reintroduce constant time comparison for hashes

  - Minor test expansion

2.009001 2015-08-08

  - Support calling Crypt::Bcrypt::Easy->crypt as a class method

2.008002 2015-08-02

  - Bytes::Random::Secure::Tiny now defaults to seeding from nonblocking
    sources; fix our constructor call when 'strong' salts are enabled

2.008001 2015-07-02

  - Switch to Bytes::Random::Secure::Tiny

2.007001 2015-06-21

  - API change; passwdcmp now returns explicit undef rather than an empty
    list

  - API change; export mkpasswd and passwdcmp by default

  - Add 'reset_seed' option to Crypt::Bcrypt::Easy's constructor as sugar for
    calling 'mkpasswd_forked' at object construction time

2.006001 2015-06-21

  - Add 'mkpasswd_forked' function for resetting salt generator seeds in child
    processes or threads.

2.005004 2015-03-11

  - 'use strictures 2;' for saner fatal warnings behavior

  - Faster tests

2.005003 2014-12-02

  - Kill constant time comparison completely; this was a silly addition that
    has only introduced bugs.

  - Switch to 'cpanfile' dependency list

2.005002 2014-09-18

  - Better Crypt::Bcrypt::Easy documentation.

  - Improvements to constant time comparison.

  - Minor optimizations & cleanups.

2.005001 2014-09-11

  - SECURITY; passwdcmp() constant time comparison fixes.
    
    The last character of a hash can be skipped during constant time
    comparison in previous versions (starting at 1.82.4).
    
    Real-world risk is reasonably low, but this does make collisions more
    likely. Mea culpa.

  - POD / test expansion

2.004002 2014-05-12

  - Minor cleanups & test improvements

2.004001 2013-10-16

  - Fix MD5 failures on machines without Crypt::Passwd::XS

2.004000 2013-10-15

  - Add mkpasswd_available function for checking avail hash methods

2.003001 2013-10-14

  - Fix failing t/04_hashopts.t on machines without SHA support

  - Use Pod::Usage to provide --help / --usage / --man

2.003000 2013-10-13

  - Support for passing in a salt generator coderef
  
  - mkpasswd() can accept parameters via a HASH

2.002000 2013-09-12

  - Use Exporter::Tiny

2.001003 2013-09-01

  - No code changes; drop prereq to perl-5.6

2.001002 2013-08-09

  - No code changes; fix 'Changes' to match CPAN::Changes::Spec

2.001001 2013-06-18

  - Kill Term::ReadKey dependency entirely; instead use POSIX::Termios to turn
    off terminal echo.  (haarg clued me in to this in a conversation he was
    having with tobyink on irc.perl.org #web-simple -- thanks!)

2.001000 2013-05-02

  - Output `bmkpasswd` prompts to STDERR for more sensible redirection

  - Make Term::ReadKey optional and warn if we don't have it

2.000003 2013-04-21

  - Cleaner --bench output

  - POD/test cleanups

2.000002 2013-04-19

  - Fix single-digit work-cost adjustment

  - Crypt::Bcrypt::Easy:
    Add ->cost() method
    Possible to pass a 'type =>' to ->crypt()
    (undocumented -- bcrypt is your friend, use it)

2.000001 2013-04-19

  - Add Crypt::Bcrypt::Easy

1.082005 2013-04-08

  - Better passwdcmp() argument-checking

  - Trap dying 'crypt()' calls in tests to avoid fuzzy fails like
    http://www.cpantesters.org/cpan/report/be021b42-9e64-11e2-8d62-e95aab8f1a3b

1.082004 2013-04-05

  - Use constant time comparison when comparing hashes.

  - POD enhancements.

1.082003 2013-04-01

  - POD tweaks

1.082002 2013-02-23

  - Lazy-build Bytes::Random::Secure instances

1.082001 2013-02-23

  - Use Bytes::Random::Secure-0.24

1.082000 2013-02-17

  - Use NonBlocking => 1 unless --strong is specified

1.081002 2013-02-17

  - Fix stray reference to deprecated HAVE_PASSWD_XS in bin/bmkpasswd

1.081001 2013-02-17
 
  - Use Crypt::Random::Seed in blocking mode to generate salts securely.

  - Switch to dzil & semantic versioning


1.07 2012-06-10

  - Slightly more sane Crypt::Passwd::XS checking.

  - Improved POD.

1.06 2012-06-09

  - Small t/ readability cleanup.

  - Missing changelog for 1.05.

1.05 2012-06-09

  - Skip MD5 tests on systems without Crypt::Passwd::XS or working MD5.

  - Fixes test failures on Windows builds without proper crypt(),
    MD5 is only included for compat with ancient hashes anyway.

1.04 2012-06-09

  - Fix a broken test on systems missing SHA support.

  - More consistent POD between bmkpasswd(1) and App::bmkpasswd

1.03 2012-06-08

  - croak() if specified work cost factor is not numeric.
  
  - croak() on unknown types.

  - Clean up an old check that is no longer relevant; have_sha() will now 
    do the Right Thing if Crypt::Passwd::XS is missing, so we don't need 
    to check if someone was stupidly futzing with package variables.

1.02 2012-06-08

  - Typo/style cleanups

1.01 2012-06-08


Build.PL
#!/usr/bin/env perl
use strict; 
use warnings;
use Module::Build;

my $build = Module::Build->new(
   module_name   => 'App::bmkpasswd',
   dist_abstract => 'bcrypt-capable mkpasswd(1) and exported helpers',
   dist_author   => 'Jon Portnoy <avenj@cobaltirc.org>',
   license => 'perl',
   
   create_makefile_pl => 'small',
   create_readme => 1,

   configure_requires => {
     'Module::Build' => '0.36',
   },
   
   build_requires => {
     'Test::More'    => 0,
   },
   
   requires => {
     'perl' => '5.8.0',

     'Crypt::Eksblowfish' => '0.003',
     
     'strictures'    => 1,

     'Getopt::Long'  => '2.24',
     'Time::HiRes'   => 0,
     
     'Try::Tiny'     => 0,
     
     'Term::ReadKey' => 0,
   },
   
   recommends => {
     'Crypt::Passwd::XS' => 0,
   },
);

$build->create_build_script;


lib/Crypt/Bcrypt/Easy.pm
package Crypt::Bcrypt::Easy;

use Carp;
use strictures 2;
use App::bmkpasswd 'mkpasswd', 'passwdcmp', 'mkpasswd_forked';

use Scalar::Util 'blessed';

use parent 'Exporter::Tiny';
our @EXPORT = 'bcrypt';
sub bcrypt {  Crypt::Bcrypt::Easy->new(@_)  }

sub DEFAULT_COST () { '08' }

=pod

=for Pod::Coverage new DEFAULT_COST

=cut

sub new {
  my ($cls, %params) = @_;
  my $cost = $params{cost} || DEFAULT_COST;
  mkpasswd_forked if $params{reset_seed};
  bless \$cost, blessed($cls) || $cls
}

sub cost { 
  my ($self) = @_;
  blessed $self ? $$self : DEFAULT_COST
}

sub compare {
  my ($self, %params) = @_;
  unless (defined $params{text} && defined $params{crypt}) {
    confess "Expected 'text =>' and 'crypt =>' params"
  }
  passwdcmp $params{text} => $params{crypt}
}

sub crypt {
  my $self = shift;
  my %params;

  if (@_ == 1) {
    $params{text} = $_[0]
  } elsif (@_ > 1) {
    %params = @_;
    confess "Expected 'text =>' param"
      unless defined $params{text};
  } else {
    confess 'Not enough arguments; expected a password'
  }

  mkpasswd $params{text} => 
    ($params{type}   || 'bcrypt'), 
    ($params{cost}   || $self->cost), 
    ($params{strong} || () )
}

1;

=pod

=head1 NAME

Crypt::Bcrypt::Easy - Simple interface to bcrypted passwords

=head1 SYNOPSIS

  use Crypt::Bcrypt::Easy;

  # Generate bcrypted passwords:
  my $plain = 'my_password';
  my $passwd = bcrypt->crypt( $plain );

  # Generate passwords with non-default options:
  my $passwd = bcrypt->crypt( text => $plain, cost => 10 );

  # Compare passwords:
  if (bcrypt->compare( text => $plain, crypt => $passwd )) {
    # Successful match
  }

  # Spawn a new instance that will generate passwords using a different
  # default workcost:
  my $bc = bcrypt( cost => 10 );
  my $passwd = $bc->crypt( $plain );

  # Without imported constructor:
  use Crypt::Bcrypt::Easy ();
  my $passwd = Crypt::Bcrypt::Easy->crypt( text => $plain, cost => 10 )

=head1 DESCRIPTION

This module provides an easy interface to creating and comparing bcrypt-hashed
passwords via L<App::bmkpasswd>'s exported helpers (which were created to
power C<bmkpasswd(1)> and are a bit awkward to use directly).

This POD briefly covers usage of this interface; see L<App::bmkpasswd> for
more details on bcrypt, internals, and documentation regarding the more
flexible functional interface.

This module uses L<Exporter::Tiny>; you can rename the L</bcrypt> function
as-needed:

  use Crypt::Bcrypt::Easy 'bcrypt' => { -as => 'bc' };

=head2 bcrypt

  my $bcrypt = bcrypt( cost => 10 );

Creates and returns a new Crypt::Bcrypt::Easy object.

The default C<cost> is '08'. This can be also be tuned for individual runs;
see L</crypt>.

(This is merely a convenience function for calling C<<
Crypt::Bcrypt::Easy->new >>.)

If your application generates passwords in multiple child processes or
threads, you can cause L<App::bmkpasswd/mkpasswd_forked> to be automatically
called during object construction in each individual process by specifying the
C<reset_seed> option:

  my $bcrypt = bcrypt( reset_seed => 1, cost => 8 );

(The C<reset_seed> option was added in C<v2.7.1>.)

=head3 crypt

Create and return a new password hash:

  my $crypted = bcrypt->crypt( 'my_password' );

Override default options (see L<App::bmkpasswd>):

  my $crypted = bcrypt->crypt(
    text   => 'my_password',
    cost   => 10,
    strong => 1,
  );

Specifying a boolean true 'strong =>' parameter enables strongly-random salts
(see L<App::bmkpasswd>).

=head3 compare

  if (bcrypt->compare(text => 'my_password', crypt => $crypted)) {
     ...
  }

Returns boolean true if hashes match. Accepts any type of hash supported by
L<App::bmkpasswd> and your system; see L<App::bmkpasswd/passwdcmp>.

=head3 cost

Returns the current work-cost value; see L<App::bmkpasswd>.

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

=cut


lib/App/bmkpasswd.pm
package App::bmkpasswd;

use strictures 2;
use Carp;
use Try::Tiny;

use Crypt::Eksblowfish::Bcrypt qw/
  bcrypt 
  en_base64
/;

use parent 'Exporter::Tiny';
our @EXPORT = qw/
  mkpasswd
  passwdcmp 
/;
our @EXPORT_OK = qw/
  mkpasswd_available
  mkpasswd_forked
/;

{ local @INC = @INC;
  pop @INC if $INC[-1] eq '.';
  require Bytes::Random::Secure::Tiny;
}
my ($brs, $brsnb);
sub get_brs {
  my (%params) = @_;
  $params{strong} ?
    $brs ||= Bytes::Random::Secure::Tiny->new(Bits => 128, NonBlocking => 0)
    : $brsnb ||= Bytes::Random::Secure::Tiny->new(Bits => 128, NonBlocking => 1)
}

sub mkpasswd_forked {
  srand; # wrt random-length salts
  undef $brs;
  undef $brsnb;
}

# can be local'd or replaced, but you get to keep both pieces ->
our $SaltGenerator = sub {
  my ($type, $strong) = @_;

  my $rnd = get_brs(strong => $strong);
  if ($type eq 'bcrypt') {
    return en_base64( $rnd->bytes(16) );
  }

  if ($type eq 'sha') {
    my $max = en_base64( $rnd->bytes(16) );
    my $initial = substr $max, 0, 8, '';
    # Drepper recommends random-length salts:
    $initial .= substr $max, 0, 1, '' for  1 .. rand 8;
    return $initial
  }

  if ($type eq 'md5') {
    return en_base64( $rnd->bytes(6) );
  }

  confess "SaltGenerator fell through, unknown type $type"
};

my %_can_haz;
sub have_passwd_xs {
  unless (defined $_can_haz{passwdxs}) {
    local @INC = @INC;
    pop @INC if $INC[-1] eq '.';
    try { require Crypt::Passwd::XS;  $_can_haz{passwdxs} = 1 } 
      catch { $_can_haz{passwdxs} = 0 };
  }
  $_can_haz{passwdxs}
}

my %_shatests = (
  sha256 => sub {
    my $testc = try { crypt('a', '$5$abc$') } catch { warn $_; () };
    $testc && index($testc, '$5$abc$') == 0 ? 1 : ()
  },
  sha512 => sub {
    my $testc = try { crypt('b', '$6$abc$') } catch { warn $_; () };
    $testc && index($testc, '$6$abc$') == 0 ? 1 : ()
  },
);

sub have_sha {
  # if we have Crypt::Passwd::XS, just use that:
  return 1 if have_passwd_xs();
  # else determine (the slow way) if SHA256/512 are available via libc:
  my $rate = $_[0] || 512;
  my $type = "sha$rate";
  return $_can_haz{$type} if defined $_can_haz{$type};
  if (exists $_shatests{$type} && $_shatests{$type}->()) {
    return $_can_haz{$type} = 1
  }
  return $_can_haz{$type} = 0
}

sub have_md5 {
  return 1 if have_passwd_xs();
  return $_can_haz{md5} if defined $_can_haz{md5};
  my $testc = try { crypt('a', '$1$abcd$') } catch { warn $_; () };
  if ($testc && index($testc, '$1$abcd$') == 0) {
    return $_can_haz{md5} = 1
  }
  return $_can_haz{md5} = 0
}


sub mkpasswd_available {
  my ($type) = @_;

  unless ($type) {
    return (
      'bcrypt',
      ( have_sha(256) ? 'sha256' : () ),
      ( have_sha(512) ? 'sha512' : () ),
      ( have_md5()    ? 'md5'    : () ),
    );
  }

  $type = lc $type;
  return 1            if $type eq 'bcrypt';
  return have_sha($1) if $type =~ /^sha-?(\d{3})$/;
  return have_md5()   if $type eq 'md5';
  return
}



sub mkpasswd {
  # mkpasswd $passwd => $type, $cost, $strongsalt;
  # mkpasswd $passwd => +{
  #   type => $type,
  #   cost => $cost,
  #   saltgen => $coderef,
  #   strong => $strongsalt,
  # }
  my $pwd = shift;
  croak "mkpasswd passed an undef password"
    unless defined $pwd;

  my %opts =
    ref $_[0] eq 'HASH' ? %{ $_[0] }
    : map {; $_ => shift } qw/type cost strong/;
  my $type = defined $opts{type} ? $opts{type} : 'bcrypt';
  my $saltgen = $opts{saltgen} || $SaltGenerator;

  my $salt;

  TYPE: {
    if ($type =~ /^bcrypt$/i) {
      my $cost = $opts{cost} || '08';

      croak 'Work cost factor must be numeric'
        unless $cost =~ /^[0-9]+$/;
      $cost = "0$cost" if length $cost == 1;

      $salt = $saltgen->(bcrypt => $opts{strong});
      my $bsettings = join '', '$2a$', $cost, '$', $salt;

      # bcrypt returns from here; everything else depends on have_passwd_xs
      return bcrypt($pwd, $bsettings)
    }

    if ($type =~ /^sha-?512$/i) {
      croak 'SHA hash requested but no SHA support available' 
        unless have_sha(512);
      $salt = join '', '$6$', $saltgen->(sha => $opts{strong}), '$';
      last TYPE
    }

    if ($type =~ /^sha(-?256)?$/i) {
      croak 'SHA hash requested but no SHA support available' 
        unless have_sha(256);
      $salt = join '', '$5$', $saltgen->(sha => $opts{strong}), '$';
      last TYPE
    }

    if ($type =~ /^md5$/i) {
      croak 'MD5 hash requested but no MD5 support available'
        unless have_md5;
      $salt = join '', '$1$', $saltgen->(md5 => $opts{strong}), '$';
      last TYPE
    }

    croak "Unknown type specified: $type"
  }

  have_passwd_xs() ?
    Crypt::Passwd::XS::crypt($pwd, $salt) : crypt($pwd, $salt)
}

sub _eq {
  my ($orig, $created) = @_;
  my $unequal = ! (length $orig == length $created);
  my $n = 0;
  no warnings 'substr';
  while ($n < length $orig) {
    my $schr = substr $created, $n, 1;
    $unequal = 1
      if substr($orig, $n, 1) ne (defined $schr ? $schr : '');
    ++$n;
  }
  ! $unequal
}

sub passwdcmp {
  my ($pwd, $crypt) = @_;
  croak 'Expected a password string and hash'
    unless defined $pwd and $crypt;

  my $pos_a = index $crypt, '$';
  my $pos_b = index $crypt, '$', 2;
  carp 'Possibly passed an invalid hash' 
    unless $pos_a == 0
    and    $pos_b == 2
    or     $pos_b == 3;

  if ($crypt =~ /^\$2a\$\d{2}\$/) {
    # Looks like bcrypt.
    return $crypt if _eq( $crypt, bcrypt($pwd, $crypt) )
  } else {
    if (have_passwd_xs) {
      return $crypt
        if _eq( $crypt, Crypt::Passwd::XS::crypt($pwd, $crypt) )
    } else {
      return $crypt
        if _eq( $crypt, crypt($pwd, $crypt) )
    }
  }

  undef
}

1;
__END__

=pod

=head1 NAME

App::bmkpasswd - bcrypt-capable mkpasswd(1) and exported bcrypt interface

=head1 SYNOPSIS

  ## From Perl:

  use App::bmkpasswd -all;
  my $bcrypted = mkpasswd($passwd);
  say 'matched' if passwdcmp($passwd, $bcrypted);

  my $stronger = mkpasswd($passwd, 'bcrypt', 12);

  ## From a shell:

  bmkpasswd --help
  
  # Generate bcrypted passwords:
  bmkpasswd

  # Defaults to work cost factor '08':
  bmkpasswd --workcost='06'

  # SHA requires Crypt::Passwd::XS or a recent libc:
  bmkpasswd --method='sha512'
  
  # Compare a hash:
  bmkpasswd --check=HASH

  # Check hash generation times:
  bmkpasswd --benchmark

=head1 DESCRIPTION

B<App::bmkpasswd> is a bcrypt-enabled C<mkpasswd> implementation.

Helper functions are also exported for use in other applications; see
L</EXPORTED> -- however L<Crypt::Bcrypt::Easy> (from this distribution)
provides an easier bcrypt-specific programmatic interface for Perl
programmers.

See C<bmkpasswd --help> for command-line usage information.

Bcrypt leverages a work-cost factor allowing hash generation
to become configurably slower as computers get faster, thereby
impeding brute-force hash generation attempts.
See L<http://codahale.com/how-to-safely-store-a-password/> for more
on why you ought to be using bcrypt or similar "adaptive" techniques.

B<SHA-256> and B<SHA-512> are supported if available. SHA support requires
either L<Crypt::Passwd::XS> or a system crypt() that can handle SHA (such as
glibc-2.7+ or modern FreeBSD builds).

This module uses L<Crypt::Eksblowfish::Bcrypt> as a back-end.

Uses L<Bytes::Random::Secure::Tiny> to generate random salts. Strongly-random salts
can also be enabled; see L</mkpasswd>.

=head1 EXPORTED

L<Crypt::Bcrypt::Easy> provides an easier programmatic interface, but only
generates bcrypt (although it can validate any supported type).  If you would
like to create crypted passwords using other methods, you can use the exported
B<mkpasswd> and B<passwdcmp> functions:

  # Import selectively:
  use App::bmkpasswd 'mkpasswd', 'passwdcmp';
  # Or import all functions:
  use App::bmkpasswd -all;

This module uses L<Exporter::Tiny> to export functions. This provides for
flexible import options. See the L<Exporter::Tiny> docs for details.

=head2 passwdcmp

Compare a password against a hash.

  if ( passwdcmp($plaintext => $crypted) ) {
    ## Successful match
  } else {
    ## Failed match
  }

B<passwdcmp> will return the hash if it is a match; otherwise, C<undef>
is returned. (This is an API change in C<v2.7.1>; prior versions return
an empty list on failure.)

As of C<v2.10>, a constant time comparison function is used to compare hashes.

=head2 mkpasswd

  my $crypted = mkpasswd($passwd);
  my $crypted = mkpasswd($passwd, $type);
  my $crypted = mkpasswd($passwd, 'bcrypt', $cost);
  my $crypted = mkpasswd($passwd, $type, $cost, $strongsalt);

  my $crypted = mkpasswd( $passwd => 
    +{
      type    => $type,
      cost    => $cost,
      strong  => $strongsalt,
      saltgen => $saltgenerator,
    }
  );

Generate hashed passwords.

By default, generates a bcrypted passwd with work-cost 08:

  $bcrypted = mkpasswd($passwd);

A different work-cost can be specified for bcrypt passwds:

  $bcrypted = mkpasswd($passwd, 'bcrypt', '10');

SHA-256 and SHA-512 are supported, in which case the work-cost value is ignored:

  $crypted = mkpasswd($passwd, 'sha256');
  $crypted = mkpasswd($passwd, 'sha512');

If a fourth boolean-true argument is specified, a strongly-random salt is
generated. This requires spare entropy, and will block if entropy-starved:

  $crypted = mkpasswd($passwd, 'bcrypt', '08', 'strong');
  $crypted = mkpasswd($passwd, 'sha512', 0, 'strong');

Options can be passed as a HASH, instead. This also lets you pass in a salt
generator coderef:

  $crypted = mkpasswd( $passwd => +{
      type => 'bcrypt',
      cost => '10',
      strong  => 0,
      saltgen => $saltgenerator,
    }
  );

The salt generator is passed the type (one of: C<bcrypt>, C<sha>, C<md5>) and
the value of the B<strong> option (default false).

  my $saltgenerator = sub {
    my ($type, $strongsalt) = @_;
    if ($type eq 'bcrypt') {
      # ...
    } elsif ($type eq 'sha') {
      # ...
    } else {
      die "Don't know how to create a salt for type '$type'!"
    }
  };

Most people want random salts, in which case the default salt generator
should be fine.

See L</mkpasswd_forked> if your application loads this module before forking
or creating threads that generate passwords.

=head2 mkpasswd_available

  my @available = mkpasswd_available;

  if ( mkpasswd_available('sha512') ) { ... }

Given no arguments, returns the list of available hash types.

Given a type (see L</mkpasswd>), returns boolean true if the method is available. ('bcrypt' is
always available.)

=head2 mkpasswd_forked

  # After a fork / new thread is created:
  mkpasswd_forked;

To retain secure salts after forking the process or creating a new thread, 
it's advisable to either only load this module after creating the new process
or call B<mkpasswd_forked> in the new process to reset the random seeds used
by salt generators.

Added in C<v2.6.1>.

=head1 AUTHOR

Jon Portnoy <jon@portnoy.me>

=for Pod::Coverage have_\w+ get_\w+

=cut
